File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.rb
# @param {String[]} tokens
# @return {Integer}
def eval_rpn(tokens)
  stack=[]
  tokens.each do |token|
    case 
    when token == '+'
      stack << (stack.pop + stack.pop)
      
    when token == '-'
      a,b = stack.pop, stack.pop
      stack << (b - a)

    when token == '*'
      stack << (stack.pop * stack.pop)

    when token == '/'
      a,b = stack.pop, stack.pop
      stack << (b.to_f / a).to_i
      
    else
      stack << token.to_i
    end
  end
  stack.last 
end

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> stack;
        for (auto& token : tokens) {
            if (token == "+") {
                int a = stack.top(); stack.pop();
                int b = stack.top(); stack.pop();
                stack.push(b + a);
            } else if (token == "-") {
                int a = stack.top(); stack.pop();
                int b = stack.top(); stack.pop();
                stack.push(b - a);
            } else  if (token == "*") {
                int a = stack.top(); stack.pop();
                int b = stack.top(); stack.pop();
                stack.push(b * a);
            } else  if (token == "/") {
                int a = stack.top(); stack.pop();
                int b = stack.top(); stack.pop();
                stack.push(b / a);
            } else {
                stack.push(stoi(token));
            }
        }
        return stack.top();
    }
};